src/lib/linux-api/src/signal.rs

```diff
@@ -0,0 +1,1197 @@
+    SEGV_ADIPERR = bindings::LINUX_SEGV_ADIPERR as i32,
+    SEGV_MTEAERR = bindings::LINUX_SEGV_MTEAERR as i32,
+    SEGV_MTESERR = bindings::LINUX_SEGV_MTESERR as i32,
+}
+
+/// Codes for SIGBUS
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeBus {
+    BUS_ADRALN = bindings::LINUX_BUS_ADRALN as i32,
+    BUS_ADRERR = bindings::LINUX_BUS_ADRERR as i32,
+    BUS_OBJERR = bindings::LINUX_BUS_OBJERR as i32,
+    BUS_MCEERR_AR = bindings::LINUX_BUS_MCEERR_AR as i32,
+    BUS_MCEERR_AO = bindings::LINUX_BUS_MCEERR_AO as i32,
+}
+
+/// Codes for SIGTRAP
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeTrap {
+    TRAP_BRKPT = bindings::LINUX_TRAP_BRKPT as i32,
+    TRAP_TRACE = bindings::LINUX_TRAP_TRACE as i32,
+    TRAP_BRANCH = bindings::LINUX_TRAP_BRANCH as i32,
+    TRAP_HWBKPT = bindings::LINUX_TRAP_HWBKPT as i32,
+    TRAP_UNK = bindings::LINUX_TRAP_UNK as i32,
+    TRAP_PERF = bindings::LINUX_TRAP_PERF as i32,
+}
+
+/// Codes for SIGIO/SIGPOLL
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodePoll {
+    POLL_IN = bindings::LINUX_POLL_IN as i32,
+    POLL_OUT = bindings::LINUX_POLL_OUT as i32,
+    POLL_MSG = bindings::LINUX_POLL_MSG as i32,
+    POLL_ERR = bindings::LINUX_POLL_ERR as i32,
+    POLL_PRI = bindings::LINUX_POLL_PRI as i32,
+    POLL_HUP = bindings::LINUX_POLL_HUP as i32,
+}
+
+/// Codes for SIGSYS
+#[allow(non_camel_case_types)]
+#[repr(i32)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+pub enum SigInfoCodeSys {
+    SYS_SECCOMP = bindings::LINUX_SYS_SECCOMP as i32,
+}
+
+#[allow(non_camel_case_types)]
+pub type linux_siginfo_t = bindings::linux_siginfo_t;
+
+pub type SigInfoDetailsKill = bindings::linux__sifields__bindgen_ty_1;
+pub type SigInfoDetailsTimer = bindings::linux__sifields__bindgen_ty_2;
+pub type SigInfoDetailsRt = bindings::linux__sifields__bindgen_ty_3;
+pub type SigInfoDetailsSigChld = bindings::linux__sifields__bindgen_ty_4;
+pub type SigInfoDetailsSigFault = bindings::linux__sifields__bindgen_ty_5;
+pub type SigInfoDetailsSigPoll = bindings::linux__sifields__bindgen_ty_6;
+pub type SigInfoDetailsSigSys = bindings::linux__sifields__bindgen_ty_7;
+
+pub enum SigInfoDetails {
+    Kill(SigInfoDetailsKill),
+    Timer(SigInfoDetailsTimer),
+    Rt(SigInfoDetailsRt),
+    SigChld(SigInfoDetailsSigChld),
+    SigFault(SigInfoDetailsSigFault),
+    SigPoll(SigInfoDetailsSigPoll),
+    SigSys(SigInfoDetailsSigSys),
+}
+
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct SigInfo(linux_siginfo_t);
+// Contains pointers, but they are understood to not necessarily be valid in the
+// current address space.
+unsafe impl VirtualAddressSpaceIndependent for SigInfo {}
```

Comment:
My view is that these types defined in the kernel API are explicitly meant to be shared between address spaces (user and kernel). While there are sometimes pointers inside of them, we don't dereference those pointers here in the wrappers. At most we we make the pointers themselves available to the caller, who should understand whether it's a pointer in the current address space or not, and needs `unsafe` to deref.

One slightly gray area is the `transmute` from pointer to `function` in `SigAction::handler`. I don't think the `transmute` itself is unsound though, and it is transmuted to an `unsafe` fn, so the caller again needs `unsafe` to actually call it.

Reply:
> My view is that these types defined in the kernel API are explicitly meant to be shared between address spaces (user and kernel).

We do share them between different user address spaces though, otherwise we wouldn't need them to be vasi. The kernel knows that any pointers given from userspace have to point to memory in the address space of the calling process, so there's no ambiguity about what address space the pointers are in. Whereas when we share pointers between user address spaces, it is ambiguous without knowing the larger context about who sets the values and when/if they should be dereferenced.

> At most we we make the pointers themselves available to the caller, who should understand whether it's a pointer in the current address space or not, and needs unsafe to deref.

Wouldn't this apply to any pointer type? What's special about this pointer, as opposed to `unsafe impl<T> VirtualAddressSpaceIndependent for *const T {}`?

Reply:
> We do share them between different user address spaces though, otherwise we wouldn't need them to be vasi.

I don't think we do? I'm thinking of shadow's address space as "kernel space". Any given instance of one of these is only going to be used in Shadow (kernel) and a single managed process that it's associated with.

> Wouldn't this apply to any pointer type? What's special about this pointer, as opposed to `unsafe impl<T> VirtualAddressSpaceIndependent for *const T {}`?

We don't provide "safe" (non-`unsafe`) methods on these types that dereference internal pointers, (partly) *because* these wrappers are designed to be `VirtualAddressSpaceIndependent`.

`unsafe impl<T> VirtualAddressSpaceIndependent for *const T {}` would be dangerous because even if the outermost pointer itself is in the current address space, it ignores whether there are any pointers inside `T` and what address space they are in.

It would particularly be a problem because it would cause the `VirtualAddressSpaceIndependent` derive macro to work in places that it shouldn't.

e.g. suppose you had:

```
struct MyHashMapWrapper {
  internal: *const HashMap<u32, u32>
}

impl MyHashMapWrapper {
  pub fn get(&self, x: &u32) -> Option<&u32> {
      unsafe { internal.ref().unwrap() }.get(x)
  }
}
```

Someone could add the line

```
#[derive(VirtualAddressSpaceIndependent)]
```

and store a `MyHashMapWrapper` in shared memory, which would make the `get` method unsound, all without any additional use of `unsafe`.

Reply:
> > We do share them between different user address spaces though, otherwise we wouldn't need them to be vasi.
> 
> I don't think we do? I'm thinking of shadow's address space as "kernel space". Any given instance of one of these is only going to be used in Shadow (kernel) and a single managed process that it's associated with.

I could see it this way, but I could also see a developer seeing it the other way too. Both shadow and the plugin are two separate processes with their own address spaces, and you're using shared memory to share the `SigInfo` between them.

> > Wouldn't this apply to any pointer type? What's special about this pointer, as opposed to `unsafe impl<T> VirtualAddressSpaceIndependent for *const T {}`?
> 
> We don't provide "safe" (non-`unsafe`) methods on these types that dereference internal pointers, (partly) _because_ these wrappers are designed to be `VirtualAddressSpaceIndependent`.

I see the `VirtualAddressSpaceIndependent` trait a little differently. For me, the `VirtualAddressSpaceIndependent` trait is a marker to say that a type is address space-independent in all cases, and anything given to you through the public API is also `VirtualAddressSpaceIndependent` (unless there's an unsafe method that has documentation that says otherwise).

With `SigInfo` (a type with the `VirtualAddressSpaceIndependent` marker), you can directly access an inner field that is not `VirtualAddressSpaceIndependent` through `sig_info.details().unwrap().l_call_addr`. In this case the `VirtualAddressSpaceIndependent` marker means "address space independent, but only as long as you don't deref any of the pointers that are given to you in the public API".

This would be like saying `UnsafeCell` should be `Sync`, because `UnsafeCell` only gives you pointers which require `unsafe` to deref. Or making a struct `Unpin` which contains a public pointer because the user code would need unsafe to dereference the pointer. In both cases we can't specify additional safety requirements on the deref since the deref is not part of the type's API.

> `unsafe impl<T> VirtualAddressSpaceIndependent for *const T {}` would be dangerous because even if the outermost pointer itself is in the current address space, it ignores whether there are any pointers inside `T` and what address space they are in.
>
> It would particularly be a problem because it would cause the VirtualAddressSpaceIndependent derive macro to work in places that it shouldn't.

But isn't that what `unsafe impl VirtualAddressSpaceIndependent for SigInfo {}` is doing? If you make a `struct Foo(SigInfo)`, it can automatically derive `VirtualAddressSpaceIndependent` even though the inner `SigInfo` has pointers which might be in another address space.

> Someone could add the line
>
> ```rust
> #[derive(VirtualAddressSpaceIndependent)]
> ```
>
> and store a MyHashMapWrapper in shared memory, which would make the get method unsound, all without any additional use of unsafe.

I'd say this is the fault of `MyHashMapWrapper`. It's doing an unsafe pointer dereference with a pointer that it doesn't know is valid. It would be like the `Foo(SigInfo)` wrapper above implementing a method to dereference a pointer in `SigInfo`.

Reply:
Ok, I'm convinced. I dropped the `VirtualAddressSpaceIndependent` markers on `SigInfo` and `SigAction`, made the accessors that can return pointers `unsafe`, and documented that we don't internally dereference pointer.